File: src/main/java/org/willy/Bench.java
package org.willy;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Fixed-depth, deterministic search over a fixed set of positions.
 * <p>
 * This exists to make the planned C++ port verifiable. Node counts are the
 * strongest available equivalence check between two implementations of the same
 * search: if the C++ engine visits exactly the same number of nodes and returns
 * the same move and score, it is almost certainly doing the same thing. Speed is
 * then a clean comparison, because both sides did identical work.
 * <p>
 * Depth is fixed rather than time, because the time-based abort in
 * {@link AIPlayer.SearchContext#check()} makes node counts depend on how fast
 * the machine happens to be, which is useless for comparison. For the same
 * reason each position gets a fresh {@link AIPlayer}: no carried-over
 * transposition table, no accumulated history heuristic, no saved memory file.
 */
class Bench {

    /**
     * Deep enough that the run does real work (~1.7M nodes) and the JVM is
     * warm by the end, shallow enough to stay under a second so it can sit in
     * CI as a regression gate.
     */
    private static final int DEFAULT_DEPTH = 8;
    private static final String DEFAULT_POSITIONS = "bench/positions.txt";

    static boolean isBenchRun(String[] args) {
        for (String a : args) if (a.startsWith("--bench")) return true;
        return false;
    }

    static void run(String[] args) {
        GameLogic.initTables();
        Zobrist.init();

        if (arg(args, "--bench-gen") != null) {
            generate(arg(args, "--positions", DEFAULT_POSITIONS));
            return;
        }
        String dump = arg(args, "--bench-dump");
        if (dump != null && !dump.isEmpty()) {
            dump(dump, arg(args, "--positions", DEFAULT_POSITIONS));
            return;
        }
        String perftDepth = arg(args, "--bench-perft");
        if (perftDepth != null) {
            perft(arg(args, "--positions", DEFAULT_POSITIONS),
                  Integer.parseInt(perftDepth.isEmpty() ? "4" : perftDepth));
            return;
        }

        int depth = Integer.parseInt(arg(args, "--depth", String.valueOf(DEFAULT_DEPTH)));
        String path = arg(args, "--positions", DEFAULT_POSITIONS);
        // Fixed time instead of fixed depth: this is the comparison that maps
        // onto play strength, since what matters in a real game is how deep an
        // engine gets within the move budget it is given.
        int movetime = Integer.parseInt(arg(args, "--movetime", "0"));

        List<String> lines;
        try {
            lines = readPositions(path);
        } catch (IOException e) {
            System.err.println("Cannot read " + path + ": " + e.getMessage());
            System.err.println("Generate it first with: --bench-gen");
            return;
        }

        if (movetime > 0) System.out.println("# fanorona bench movetime=" + movetime + "ms positions=" + lines.size());
        else System.out.println("# fanorona bench depth=" + depth + " positions=" + lines.size());
        long totalNodes = 0, totalHits = 0, totalMs = 0, totalDepth = 0;

        for (int i = 0; i < lines.size(); i++) {
            GameState s = GameStateCodec.fromText(lines.get(i));

            SearchConfig cfg = new SearchConfig();
            cfg.maxDepth = movetime > 0 ? 1000 : depth;
            cfg.timeLimitMs = movetime > 0 ? movetime : Integer.MAX_VALUE;
            cfg.logger = null;
            AIPlayer ai = new AIPlayer(cfg);

            long t0 = System.nanoTime();
            AIPlayer.AIResult r = ai.think(s);
            long ms = (System.nanoTime() - t0) / 1_000_000;

            if (movetime > 0)
                System.out.printf("%d depth=%d nodes=%d score=%d bestMove=%d%n",
                        i, r.depth, r.nodes, r.score, r.bestMove);
            else
                System.out.printf("%d nodes=%d ttHits=%d score=%d bestMove=%d%n",
                        i, r.nodes, r.ttHits, r.score, r.bestMove);
            totalNodes += r.nodes;
            totalHits += r.ttHits;
            totalDepth += r.depth;
            totalMs += ms;
        }
        if (movetime > 0) System.out.printf("# total depth=%d nodes=%d%n", totalDepth, totalNodes);
        else

        System.out.printf("# total nodes=%d ttHits=%d%n", totalNodes, totalHits);

        // Timing goes to stderr so stdout stays byte-stable and can be diffed
        // straight against the golden file:
        //     --bench > bench/golden.txt
        System.err.printf("# wall %d ms, %.0f knps%n",
                totalMs, totalMs > 0 ? totalNodes / (double) totalMs : 0.0);
    }

    /** Skips blank lines and {@code #} comments so the file can document itself. */
    static List<String> readPositions(String path) throws IOException {
        List<String> out = new ArrayList<>();
        for (String line : Files.readAllLines(Path.of(path), StandardCharsets.UTF_8)) {
            String t = line.trim();
            if (!t.isEmpty() && !t.startsWith("#")) out.add(t);
        }
        return out;
    }

    // ---------------------------------------------------------------- perft

    /**
     * Counts leaf nodes of the pure move-generation tree.
     * <p>
     * Unlike the search benchmark, this touches no transposition table, no
     * evaluation and no move ordering — only the rules. That makes it the one
     * correctness check that stays valid through every search optimisation:
     * a faster transposition table or a parallel search is *supposed* to change
     * search node counts, but it must never change perft.
     */
    static long perft(GameState s, int depth) {
        List<Move> moves = GameLogic.getDetailedMoves(s);
        if (depth <= 1) return moves.size();
        long n = 0;
        for (Move m : moves) n += perft(GameLogic.step(s, m.actionId).state, depth - 1);
        return n;
    }

    private static void perft(String path, int maxDepth) {
        List<String> lines = positions(path);
        System.out.println("# perft maxDepth=" + maxDepth + " positions=" + lines.size());
        for (int i = 0; i < lines.size(); i++) {
            GameState s = GameStateCodec.fromText(lines.get(i));
            StringBuilder sb = new StringBuilder();
            sb.append(i);
            for (int d = 1; d <= maxDepth; d++) sb.append(' ').append(perft(s, d));
            System.out.println(sb);
        }
    }

    // ---------------------------------------------------------------- sub-harnesses

    /**
     * Dumps one layer of the engine at a time.
     * <p>
     * A whole-search node-count divergence is close to impossible to localise,
     * so the C++ port is checked bottom-up: get {@code zobrist} matching before
     * looking at {@code movegen}, {@code movegen} before {@code step}, and so on.
     * Each mode prints a stable, diffable text stream and nothing else.
     */
    private static void dump(String what, String path) {
        switch (what) {
            case "zobrist" -> {
                // Depends on Java's exact LCG; the first thing to break in a port.
                for (int i = 0; i < GameLogic.NUM_POS; i++)
                    System.out.printf("P %d 0 %016x%nP %d 1 %016x%n",
                            i, Zobrist.P[i][0], i, Zobrist.P[i][1]);
                System.out.printf("T %016x%n", Zobrist.T);
            }
            case "codec" -> {
                int bad = 0;
                for (String line : positions(path)) {
                    GameState s = GameStateCodec.fromText(line);
                    String back = GameStateCodec.toText(s);
                    if (!back.equals(line)) {
                        bad++;
                        System.out.println("MISMATCH\n  in : " + line + "\n  out: " + back);
                    }
                }
                System.out.println(bad == 0 ? "# codec round-trip OK" : "# " + bad + " MISMATCHES");
            }
            case "movegen" -> {
                for (String line : positions(path)) {
                    GameState s = GameStateCodec.fromText(line);
                    System.out.printf("pos %016x%n", s.zobristHash);
                    for (Move m : GameLogic.getDetailedMoves(s))
                        System.out.printf("  move id=%d from=%d to=%d type=%s victims=%d%n",
                                m.actionId, m.from, m.to, m.type, m.victims.size());
                }
            }
            case "step" -> {
                for (String line : positions(path)) {
                    GameState s = GameStateCodec.fromText(line);
                    System.out.printf("pos %016x%n", s.zobristHash);
                    for (Move m : GameLogic.getDetailedMoves(s)) {
                        GameLogic.StepResult r = GameLogic.step(s, m.actionId);
                        System.out.printf("  %d -> %s win=%b hash=%016x lastDir=%d%n",
                                m.actionId, GameStateCodec.toText(r.state), r.win,
                                r.state.zobristHash, r.state.lastDir);
                    }
                }
            }
            case "eval" -> {
                for (String line : positions(path)) {
                    GameState s = GameStateCodec.fromText(line);
                    System.out.printf("%016x %d%n", s.zobristHash, AIPlayer.evaluate(s));
                    // Also evaluate every child, for far more coverage than the
                    // six root positions alone would give.
                    for (Move m : GameLogic.getDetailedMoves(s)) {
                        GameState c = GameLogic.step(s, m.actionId).state;
                        System.out.printf("  %d %016x %d%n", m.actionId, c.zobristHash, AIPlayer.evaluate(c));
                    }
                }
            }
            default -> System.err.println(
                    "unknown dump '" + what + "'; expected zobrist|codec|movegen|step|eval");
        }
    }

    private static List<String> positions(String path) {
        try {
            return readPositions(path);
        } catch (IOException e) {
            throw new RuntimeException("Cannot read " + path + " (generate it with --bench-gen)", e);
        }
    }

    // ---------------------------------------------------------------- generation

    private static final String START_BOARD =
            "wwwwwwwwwwwwwwwwww" + "wbwb.wbwb" + "bbbbbbbbbbbbbbbbbb";

    /**
     * Walks one deterministic game and saves the positions worth benchmarking.
     * <p>
     * Move choice is greedy on captures with an actionId tie-break, so the walk
     * is reproducible without a random seed, and it naturally passes through
     * capture chains — which is what exercises the combo branch in
     * {@code GameLogic.step} and the {@code d <= 0 && inCombo} extension in the
     * search, the two places a port is most likely to diverge.
     */
    private static void generate(String path) {
        // Play the whole game first, then sample it proportionally. Sampling at
        // fixed ply numbers would miss: greedy capture play finishes in far fewer
        // moves than a real game, so most fixed offsets fall past the end.
        List<GameState> game = new ArrayList<>();
        GameState s = GameStateCodec.fromText(START_BOARD + " 1 0 -1 -1 0");
        game.add(s);

        boolean ended = false;
        for (int ply = 1; ply <= 400; ply++) {
            List<Move> moves = GameLogic.getDetailedMoves(s);
            if (moves.isEmpty()) {
                ended = true;
                break;
            }
            Move best = moves.stream()
                    .filter(m -> m.actionId != 720)
                    .max(Comparator.<Move>comparingInt(m -> m.victims.size())
                            .thenComparing(m -> -m.actionId))
                    .orElse(moves.get(0));

            GameLogic.StepResult res = GameLogic.step(s, best.actionId);
            s = res.state;
            game.add(s);
            if (res.win) {
                ended = true;
                break;
            }
        }

        List<String> picked = new ArrayList<>();
        List<String> notes = new ArrayList<>();
        int n = game.size();

        add(picked, notes, game.get(0), "opening (initial position)");
        add(picked, notes, game.get(Math.max(1, n / 6)), "early game");
        add(picked, notes, game.get(n / 2), "midgame");
        add(picked, notes, game.get(Math.min(n - 1, n * 4 / 5)), "late game");

        // The combo branch in step() and the `d <= 0 && inCombo` extension in the
        // search are where a port is most likely to diverge, so make sure at
        // least one mid-chain position is covered.
        for (GameState g : game) {
            if (g.inCombo) {
                add(picked, notes, g, "mid-capture chain (inCombo)");
                break;
            }
        }
        if (ended) add(picked, notes, game.get(n - 1),
                "terminal (game over: no moves, or opponent wiped out)");

        StringBuilder sb = new StringBuilder();
        sb.append("# Benchmark positions for the Java/C++ parity check.\n");
        sb.append("# Format: <45 board chars .wb> <player> <inCombo> <comboPiece> <prevPos> <visitedHex>\n");
        sb.append("# Regenerate with: --bench-gen\n");
        for (int i = 0; i < picked.size(); i++) {
            sb.append("# ").append(i).append(": ").append(notes.get(i)).append('\n');
            sb.append(picked.get(i)).append('\n');
        }
        try {
            Path p = Path.of(path);
            if (p.getParent() != null) Files.createDirectories(p.getParent());
            Files.writeString(p, sb.toString(), StandardCharsets.UTF_8);
            System.out.println("wrote " + path + " (" + picked.size() + " positions)");
        } catch (IOException e) {
            System.err.println("Cannot write " + path + ": " + e.getMessage());
        }
    }

    private static void add(List<String> picked, List<String> notes, GameState s, String note) {
        picked.add(GameStateCodec.toText(s));
        notes.add(note);
    }

    // ---------------------------------------------------------------- arg helpers

    private static String arg(String[] args, String name) {
        for (String a : args) {
            if (a.equals(name)) return "";
            if (a.startsWith(name + "=")) return a.substring(name.length() + 1);
        }
        return null;
    }

    private static String arg(String[] args, String name, String fallback) {
        String v = arg(args, name);
        return (v == null || v.isEmpty()) ? fallback : v;
    }
}

File: engine/src/bench.cpp
// Mirror of Bench.java. Output on stdout must be byte-identical to the Java
// side's, so the parity check is a plain diff against bench/golden.txt.
#include "fanorona.h"

#include <chrono>
#include <cinttypes>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <iostream>
#include <limits>
#include <random>
#include <utility>
#include <string>
#include <vector>

using namespace fanorona;

namespace {

constexpr int DEFAULT_DEPTH = 8;
constexpr const char* DEFAULT_POSITIONS = "bench/positions.txt";

std::vector<std::string> read_positions(const std::string& path) {
    std::ifstream in(path);
    if (!in) {
        std::fprintf(stderr, "Cannot read %s (generate it with the Java --bench-gen)\n", path.c_str());
        std::exit(2);
    }
    std::vector<std::string> out;
    std::string line;
    while (std::getline(in, line)) {
        if (!line.empty() && line.back() == '\r') line.pop_back();
        size_t b = line.find_first_not_of(" \t");
        if (b == std::string::npos) continue;
        size_t e = line.find_last_not_of(" \t");
        std::string t = line.substr(b, e - b + 1);
        if (t.empty() || t[0] == '#') continue;
        out.push_back(t);
    }
    return out;
}

int64_t now_ms() {
    using namespace std::chrono;
    return duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
}

const char* arg_value(int argc, char** argv, const char* name) {
    size_t n = std::strlen(name);
    for (int i = 1; i < argc; i++) {
        if (std::strncmp(argv[i], name, n) == 0) {
            if (argv[i][n] == '=') return argv[i] + n + 1;
            if (argv[i][n] == '\0') return "";
        }
    }
    return nullptr;
}

// Counts leaf nodes of the pure move-generation tree: no transposition table,
// no evaluation, no ordering, only the rules. This is the correctness check
// that survives every search optimisation -- a faster table or a parallel
// search is meant to change search node counts, but must never change perft.
int64_t perft(const GameState& s, int depth) {
    std::vector<Move> moves = get_detailed_moves(s);
    if (depth <= 1) return static_cast<int64_t>(moves.size());
    int64_t n = 0;
    for (const Move& m : moves) n += perft(step(s, m.actionId).state, depth - 1);
    return n;
}

void run_perft(const std::string& path, int maxDepth) {
    auto positions = read_positions(path);
    std::printf("# perft maxDepth=%d positions=%zu\n", maxDepth, positions.size());
    for (size_t i = 0; i < positions.size(); i++) {
        GameState s = from_text(positions[i]);
        std::printf("%zu", i);
        for (int d = 1; d <= maxDepth; d++) std::printf(" %" PRId64, perft(s, d));
        std::printf("\n");
    }
}

// Plays two engine configurations against each other.
//
// Needed because "this evaluation change should make it stronger" is a
// hypothesis, not a result. Node counts and nanoseconds cannot answer it --
// only games can. Openings are randomised and colours alternate so neither
// side is judged on a single line of play.
struct MatchSide {
    int huntCap;   // eval_hunt_cap for this side; negative = legacy un-normalised sum
    const char* name;
};

struct GameOutcome {
    int winner = 0;      // +1 white, -1 black, 0 unresolved
    int materialWhite = 0, materialBlack = 0;
};

GameOutcome play_game(const MatchSide& white, const MatchSide& black, int movetimeMs,
                      uint64_t seed, int openingPlies, int plyCap) {
    GameState s = from_text(
        "wwwwwwwwwwwwwwwwww" "wbwb.wbwb" "bbbbbbbbbbbbbbbbbb" " 1 0 -1 -1 0");

    SearchConfig cfg;
    cfg.timeLimitMs = movetimeMs;
    cfg.maxDepth = 1000;
    cfg.threads = 1;
    cfg.hashMB = 32;
    AIPlayer aiW(cfg), aiB(cfg);

    std::mt19937_64 rng(seed);
    Move buf[MAX_MOVES];

    auto material = [](const GameState& g) {
        // Bitboards are side-to-move relative; report them as white/black.
        int mine = std::popcount(g.myPieces), theirs = std::popcount(g.oppPieces);
        return g.player == 1 ? std::pair<int, int>{mine, theirs}
                             : std::pair<int, int>{theirs, mine};
    };

    for (int ply = 0; ply < plyCap; ply++) {
        int n = generate_moves(s, buf);
        if (n == 0) {  // side to move has no reply: it loses
            auto [w, b] = material(s);
            return {-s.player, w, b};
        }

        int action;
        if (ply < openingPlies) {
            action = buf[rng() % n].actionId;      // randomised opening
        } else {
            const MatchSide& side = (s.player == 1) ? white : black;
            eval_hunt_cap = side.huntCap;          // the setting under test
            AIPlayer& ai = (s.player == 1) ? aiW : aiB;
            action = ai.think(s).bestMove;
            bool legal = false;
            for (int i = 0; i < n; i++) if (buf[i].actionId == action) legal = true;
            if (!legal) action = buf[0].actionId;
        }
        StepResult r = step(s, action);
        if (r.win) {
            auto [w, b] = material(r.state);
            return {r.state.player, w, b};         // mover wiped the opponent out
        }
        s = r.state;
    }
    // Fanorona draws a lot, and at these time controls most games hit the cap.
    // Reporting them all as "draw" throws away the information that one side
    // was up eight pieces, so unresolved games are adjudicated on material.
    auto [w, b] = material(s);
    return {0, w, b};
}

void run_match(int games, int movetimeMs, int huntA, int huntB, int openingPlies, int plyCap) {
    MatchSide A{huntA, "A"}, B{huntB, "B"};
    int winA = 0, winB = 0, draw = 0, adjA = 0, adjB = 0;
    long long marginSum = 0;
    std::printf("# match: A(huntcap=%d) vs B(huntcap=%d), %d games, %dms/move, "
                "%d random opening plies, %d ply cap\n",
                huntA, huntB, games, movetimeMs, openingPlies, plyCap);
    for (int g = 0; g < games; g++) {
        // Same seed for the pair with colours swapped, so both configurations
        // play the identical opening from both sides and an easy start cannot
        // flatter whichever engine happened to draw it.
        bool aIsWhite = (g % 2) == 0;
        uint64_t seed = 0x9E3779B97F4A7C15ULL * (g / 2 + 1);
        GameOutcome o = aIsWhite ? play_game(A, B, movetimeMs, seed, openingPlies, plyCap)
                                 : play_game(B, A, movetimeMs, seed, openingPlies, plyCap);
        int aWinner = aIsWhite ? o.winner : -o.winner;
        int aMargin = aIsWhite ? (o.materialWhite - o.materialBlack)
                               : (o.materialBlack - o.materialWhite);
        marginSum += aMargin;
        const char* label;
        if (aWinner > 0)      { winA++; label = "A wins"; }
        else if (aWinner < 0) { winB++; label = "B wins"; }
        else if (aMargin > 0) { adjA++; draw++; label = "cap, A ahead"; }
        else if (aMargin < 0) { adjB++; draw++; label = "cap, B ahead"; }
        else                  { draw++;        label = "cap, level"; }
        std::printf("  game %2d: %-13s (material %+d)\n", g + 1, label, aMargin);
        std::fflush(stdout);
    }
    double pct = games ? (winA + 0.5 * draw) * 100.0 / games : 0.0;
    std::printf("# decisive: A %d, B %d | capped: %d (A ahead %d, B ahead %d)\n",
                winA, winB, draw, adjA, adjB);
    std::printf("# A scores %.1f%% | mean material margin %+.2f\n",
                pct, games ? (double)marginSum / games : 0.0);
}

void dump(const std::string& what, const std::string& path) {
    if (what == "zobrist") {
        for (int i = 0; i < NUM_POS; i++) {
            std::printf("P %d 0 %016llx\n", i, (unsigned long long)zobrist::P[i][0]);
            std::printf("P %d 1 %016llx\n", i, (unsigned long long)zobrist::P[i][1]);
        }
        std::printf("T %016llx\n", (unsigned long long)zobrist::T);
        return;
    }
    auto positions = read_positions(path);
    if (what == "codec") {
        int bad = 0;
        for (const auto& line : positions) {
            GameState s = from_text(line);
            std::string back = to_text(s);
            if (back != line) {
                bad++;
                std::printf("MISMATCH\n  in : %s\n  out: %s\n", line.c_str(), back.c_str());
            }
        }
        std::printf(bad == 0 ? "# codec round-trip OK\n" : "# %d MISMATCHES\n", bad);
    } else if (what == "movegen") {
        for (const auto& line : positions) {
            GameState s = from_text(line);
            std::printf("pos %016llx\n", (unsigned long long)s.zobristHash);
            for (const Move& m : get_detailed_moves(s))
                std::printf("  move id=%d from=%d to=%d type=%s victims=%d\n",
                            m.actionId, m.from, m.to, to_string(m.type), m.victims());
        }
    } else if (what == "step") {
        for (const auto& line : positions) {
            GameState s = from_text(line);
            std::printf("pos %016llx\n", (unsigned long long)s.zobristHash);
            for (const Move& m : get_detailed_moves(s)) {
                StepResult r = step(s, m.actionId);
                std::printf("  %d -> %s win=%s hash=%016llx lastDir=%d\n",
                            m.actionId, to_text(r.state).c_str(), r.win ? "true" : "false",
                            (unsigned long long)r.state.zobristHash, r.state.lastDir);
            }
        }
    } else if (what == "eval") {
        for (const auto& line : positions) {
            GameState s = from_text(line);
            std::printf("%016llx %d\n", (unsigned long long)s.zobristHash, evaluate(s));
            for (const Move& m : get_detailed_moves(s)) {
                GameState c = step(s, m.actionId).state;
                std::printf("  %d %016llx %d\n", m.actionId,
                            (unsigned long long)c.zobristHash, evaluate(c));
            }
        }
    } else {
        std::fprintf(stderr, "unknown dump '%s'; expected zobrist|codec|movegen|step|eval\n",
                     what.c_str());
    }
}

}  // namespace

int main(int argc, char** argv) {
    init_tables();
    zobrist::init();
    bool forceScalar = false;
    if (const char* e = arg_value(argc, argv, "--eval"); e && std::strcmp(e, "scalar") == 0) forceScalar = true;
    init_eval(forceScalar);
    if (const char* h = arg_value(argc, argv, "--hunt-cap"); h && *h) eval_hunt_cap = std::atoi(h);

    std::string path = DEFAULT_POSITIONS;
    if (const char* p = arg_value(argc, argv, "--positions"); p && *p) path = p;

    if (const char* d = arg_value(argc, argv, "--bench-dump"); d && *d) {
        dump(d, path);
        return 0;
    }
    if (const char* d = arg_value(argc, argv, "--bench-perft"); d) {
        run_perft(path, (*d) ? std::atoi(d) : 4);
        return 0;
    }
    if (arg_value(argc, argv, "--match")) {
        auto num = [&](const char* n, int def) {
            const char* v = arg_value(argc, argv, n);
            return (v && *v) ? std::atoi(v) : def;
        };
        run_match(num("--games", 20), num("--movetime", 100), num("--huntA", 50),
                  num("--huntB", -1), num("--opening", 6), num("--plycap", 200));
        return 0;
    }

    int depth = DEFAULT_DEPTH;
    if (const char* d = arg_value(argc, argv, "--depth"); d && *d) depth = std::atoi(d);
    int threads = 1;
    if (const char* t = arg_value(argc, argv, "--threads"); t && *t) threads = std::atoi(t);
    int hashMB = 256;
    if (const char* h = arg_value(argc, argv, "--hash"); h && *h) hashMB = std::atoi(h);

    // Fixed time instead of fixed depth: this is the comparison that maps onto
    // play strength, since what matters in a game is how deep an engine gets
    // within the move budget it is given.
    int movetime = 0;
    if (const char* m = arg_value(argc, argv, "--movetime"); m && *m) movetime = std::atoi(m);

    auto positions = read_positions(path);
    if (movetime > 0)
        std::printf("# fanorona bench movetime=%dms positions=%zu\n", movetime, positions.size());
    else
        std::printf("# fanorona bench depth=%d positions=%zu\n", depth, positions.size());
    if (threads != 1 || hashMB != 256 || movetime > 0)
        std::fprintf(stderr, "# threads=%d hash=%dMB eval=%s\n", threads, hashMB, eval_backend_name());

    int64_t totalNodes = 0, totalHits = 0, totalMs = 0, totalDepth = 0;
    for (size_t i = 0; i < positions.size(); i++) {
        GameState s = from_text(positions[i]);

        SearchConfig cfg;
        cfg.maxDepth = movetime > 0 ? 1000 : depth;
        cfg.timeLimitMs = movetime > 0 ? movetime : std::numeric_limits<int32_t>::max();
        cfg.threads = threads;
        cfg.hashMB = hashMB;
        AIPlayer ai(cfg);

        int64_t t0 = now_ms();
        SearchResult r = ai.think(s);
        totalMs += now_ms() - t0;

        if (movetime > 0)
            std::printf("%zu depth=%d nodes=%" PRId64 " score=%d bestMove=%d\n",
                        i, r.depth, r.nodes, r.score, r.bestMove);
        else
            std::printf("%zu nodes=%" PRId64 " ttHits=%" PRId64 " score=%d bestMove=%d\n",
                        i, r.nodes, r.ttHits, r.score, r.bestMove);
        totalNodes += r.nodes;
        totalHits += r.ttHits;
        totalDepth += r.depth;
    }
    if (movetime > 0)
        std::printf("# total depth=%" PRId64 " nodes=%" PRId64 "\n", totalDepth, totalNodes);
    else
        std::printf("# total nodes=%" PRId64 " ttHits=%" PRId64 "\n", totalNodes, totalHits);

    // Timing to stderr so stdout stays diffable against the golden file.
    std::fprintf(stderr, "# wall %" PRId64 " ms, %.0f knps\n", totalMs,
                 totalMs > 0 ? totalNodes / (double)totalMs : 0.0);
    return 0;
}
